### ISSUE TITLE
Chineses Messy Code

### ISSUE THREAD

**Issue by wyc7** [mitre/caldera#2119]
Hi, I recently discovered that Chinese garbled characters appear in the command output. How can I solve it? Can you help me?

Screenshots：
![image](https://user-images.githubusercontent.com/48666541/114575462-c5b1bd80-9cac-11eb-9031-492930ecddb6.png)
![image](https://user-images.githubusercontent.com/48666541/114575513-d3ffd980-9cac-11eb-9645-b7adcb07ef04.png)
![image](https://user-images.githubusercontent.com/48666541/114576069-61dbc480-9cad-11eb-91d1-11d0a1b025f8.png)

**Comment by garunagiri** [comment]
Hi wyc7,

Currently, CALDERA Base64 encodes all command output for storage, and decodes it when presenting it to the user. Unfortunately, Chinese characters and potentially other scripts are not supported at this time, as they are not uniformly processed by shells on the target host. This in turn means that the results cannot be properly processed by the CALDERA server during Base64 encoding and decoding.

**Comment by garunagiri** [comment]
Hy wyc7,

We're digging a little deeper into the issue, and would like to get some more information on this.

Can you show us a screenshot of the `Get-ChildItem` command output when you run it in a native PowerShell session please?

Additionally, can you paste the specific characters that are causing the trouble? We'd like to perform some additional testing in order to better understand the issue.

**Comment by wyc7** [comment]
Hi!
Run in the native psh, the command output is normal.
Screenshot：
![image](https://user-images.githubusercontent.com/48666541/115106955-744e4a80-9f9a-11eb-9cbe-bc4e36a87896.png)

Then, I found that as long as the command output is in Chinese, there are garbled characters.

**Comment by wyc7** [comment]
Additionally,Are you debugging?I found that if it is a PNG image with a Chinese name, Caldera cannot successfully copy it to the staged directory.
Screenshots：
PNG image with Chinese name：
![image](https://user-images.githubusercontent.com/48666541/115108084-24bf4d00-9fa1-11eb-9e38-0f806d27b359.png)
![image](https://user-images.githubusercontent.com/48666541/115108087-2a1c9780-9fa1-11eb-8076-612dcc50adc0.png)

PNG image without Chinese name:
![image](https://user-images.githubusercontent.com/48666541/115108156-710a8d00-9fa1-11eb-8faf-78e1a04da315.png)
![image](https://user-images.githubusercontent.com/48666541/115108163-7ec01280-9fa1-11eb-9eee-f3f35563474c.png)

**Comment by wyc7** [comment]
Additionally,When I run the 'super spy' adversary profiles, the following 4 abilities will be garbled:
Screenshots：
![image](https://user-images.githubusercontent.com/48666541/115108472-3570c280-9fa3-11eb-8d43-e4d5d684c939.png)
1.Discover antivirus programs:
![image](https://user-images.githubusercontent.com/48666541/115108479-3f92c100-9fa3-11eb-958e-98a19125afcf.png)
2.Scan WIFI networks
![image](https://user-images.githubusercontent.com/48666541/115108482-49b4bf80-9fa3-11eb-878c-3daf2665acfa.png)
3.Preferred WIFI
![image](https://user-images.githubusercontent.com/48666541/115108509-77016d80-9fa3-11eb-864a-08e2649b5d6c.png)
4.Sniff network traffic
![image](https://user-images.githubusercontent.com/48666541/115108524-87b1e380-9fa3-11eb-89f2-5d4a7b736ab8.png)

At the same time, the shell terminal will report an error.
Screenshot：
![image](https://user-images.githubusercontent.com/48666541/115108567-c8a9f800-9fa3-11eb-91dd-ea641a620930.png)

**Comment by garunagiri** [comment]
Thank you for all of that information! In order to best debug this, I need to reproduce the issue. What is the method you used to display Chinese characters within PowerShell? Did you install a new font? Or did you change the encoding that PowerShell uses?

**Comment by wyc7** [comment]
I didn't do anything, because I am in China now and my win10 operating system is in Chinese.

**Comment by garunagiri** [comment]
We've made progress on the garbled characters on the front end issue, and a fix should be available shortly.

The terminal error seems to be a slightly different issue however, and so far I've been unable to reproduce it.
To help me understand what's going on, can you manually run `.\obfuscated_payload.ps1 -Scan` in a PowerShell window from a directory that does **not** contain `obfuscated_payload.ps1`, and paste the results here as a screenshot please?

**Comment by wyc7** [comment]
![image](https://user-images.githubusercontent.com/48666541/115487274-f2f10380-a28a-11eb-9b45-6c71e739a3d2.png)

**Comment by garunagiri** [comment]
Thank you for the image wyc7.

I dug deeper into this issue, and was able to reproduce it with a Windows 10 VM installed with Chinese (Simplified) as the default language. PowerShell, and the Windows console as a whole, uses different codecs to display text in different languages. There unfortunately is not a clean way to include all codecs in Python in CALDERA, and so I have to revert back to my original comment, in that PowerShell sessions that display messages in Chinese by default are not supported for the moment.

There will be support however for Chinese characters that are represented using UTF-8, coming shortly. Thank you for bringing that to our attention!

**Comment by wyc7** [comment]
Oh, I really look forward to using the new version of caldera that supports Chinese characters! Thanks in advance!

**Comment by garunagiri** [comment]
We just merged in the changes for UTF-8 Chinese (and others) character support here:
https://github.com/mitre/caldera/pull/2130

I hope it helps!

**Comment by Difficultcl** [comment]
Thank you for your method, but after the modification, the garbled code still exists.
![image](https://user-images.githubusercontent.com/38584419/115844677-a6582480-a452-11eb-93fe-0746bcdcf75c.png)
![image](https://user-images.githubusercontent.com/38584419/115844704-ad7f3280-a452-11eb-85c6-6122e545b32d.png)
½Ó¿Ú: 192.168.10.137 --- 0x3
  Internet µØÖ·         ÎïÀíµØÖ·              ÀàÐÍ
  192.168.10.135        00-0c-29-15-ea-dc     ¶¯Ì¬        
  192.168.10.255        ff-ff-ff-ff-ff-ff     ¾²Ì¬        
  224.0.0.22            01-00-5e-00-00-16     ¾²Ì¬        
  224.0.0.251           01-00-5e-00-00-fb     ¾²Ì¬        
  239.255.255.250       01-00-5e-7f-ff-fa     ¾²Ì¬        
接口: 192.168.10.137 --- 0x3
  Internet 地址         物理地址              类型
  192.168.10.135        00-0c-29-15-ea-dc     动态
  192.168.10.255        ff-ff-ff-ff-ff-ff     静态
  224.0.0.22            01-00-5e-00-00-16     静态
  224.0.0.251           01-00-5e-00-00-fb     静态
  239.255.255.250       01-00-5e-7f-ff-fa     静态

After I change the language of windows 10 to English, CALDERA can convert the output of command to fact.
![image](https://user-images.githubusercontent.com/38584419/115844879-de5f6780-a452-11eb-83d8-7216c9d919b2.png)
![image](https://user-images.githubusercontent.com/38584419/115844916-e5867580-a452-11eb-9126-5e7c9424f2a4.png)

**Comment by garunagiri** [comment]
Hi Difficultcl,

Yes, unfortunately, as I mentioned in a previous comment, PowerShell uses a different character encoding than UTF-8 when its language is changed. CALDERA currently does not support other encodings.

The change we made was to support characters from other scripts that are encoded in UTF-8. So if your PowerShell language was in English, but used a font that supported Chinese characters such as NSimSun, then the output would be shown correctly within CALDERA.

I apologize that the language issue couldn't be solved at this moment, but I hope the change we made is still useful to you.
